### ISSUE TITLE
Add LDA example to illustrate some usage of `flax_module`/`haiku_module`

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/numpyro#1033]
After #1016 and #989, it would be nice to port [ProdLDA](http://pyro.ai/examples/prodlda.html#ProdLDA:-Latent-Dirichlet-Allocation-with-Product-of-Experts) example in Pyro to illustrate some usage of JAX nn frameworks `flax`/`haiku` in NumPyro. We might consider replacing `SoftmaxNormal` there with `Dirichlet` distribution. This is a good first issue for those who are interested in both NLP and Bayesian inference.

**Comment by tcbegley** [comment]
I'm interested in this and happy to start working on it.

**Comment by tcbegley** [comment]
Hi @fehiepsi, I have a question about the Pyro version of this notebook.

Both encoder and decoder have dropout layers, is there a way to know inside the model / guide if they are being run in a training context by e.g. `SVI` versus being used in a generative / predictive context? In the latter case I think you would want to ensure that dropout is turned off right?

From what I can tell in the Pyro / ProdLDA example it doesn't really matter, because the encoder / decoder only ever do a forward pass during training / inference and then the weight vector of the decoder are directly extracted once training is complete. Is that right?

**Comment by fehiepsi** [comment]
You are right, only the weight vector is important to extract topics' word distributions. To let `model/guide` know about training context, you can add an argument `training=True/False` to them:
```python
def Decoder(x, training=...):
    dropout_rate = ... if training else 0.
    ...

def model(..., training=...):
    decoder = haiku_module("encoder", hk.transform(Decoder), apply_rng=True, ...)
    out = decoder(numpyro.prng_key(), x, False)
```

**Comment by fehiepsi** [comment]
Thanks, @tcbegley!
